rasa/utils/train_utils.py

```diff
@@ -0,0 +1,928 @@
+    elif evaluate_on_num_examples < len(label_counts):
+        raise ValueError(
+            "Validation set of {} is too small. It should be "
+            "at least equal to number of classes {}."
+            "".format(evaluate_on_num_examples, len(label_counts))
+        )
+
+    counts = np.array([label_counts[label] for label in session_data.labels])
+
+    multi_X = session_data.X[counts > 1]
+    multi_Y = session_data.Y[counts > 1]
+    multi_labels = session_data.labels[counts > 1]
+
+    solo_X = session_data.X[counts == 1]
+    solo_Y = session_data.Y[counts == 1]
+    solo_labels = session_data.labels[counts == 1]
+
+    (X_train, X_val, Y_train, Y_val, labels_train, labels_val) = train_test_split(
+        multi_X,
+        multi_Y,
+        multi_labels,
+        test_size=evaluate_on_num_examples,
+        random_state=random_seed,
+        stratify=multi_labels,
+    )
+    X_train = np.concatenate([X_train, solo_X])
+    Y_train = np.concatenate([Y_train, solo_Y])
+    labels_train = np.concatenate([labels_train, solo_labels])
+
+    return (
+        SessionData(X=X_train, Y=Y_train, labels=labels_train),
+        SessionData(X=X_val, Y=Y_val, labels=labels_val),
+    )
+
+
+def shuffle_session_data(session_data: "SessionData") -> "SessionData":
+    """Shuffle session data."""
+
+    ids = np.random.permutation(len(session_data.X))
+    return SessionData(
+        X=session_data.X[ids], Y=session_data.Y[ids], labels=session_data.labels[ids]
+    )
+
+
+# noinspection PyPep8Naming
+def balance_session_data(
+    session_data: "SessionData", batch_size: int, shuffle: bool
+) -> "SessionData":
+    """Mix session data to account for class imbalance."""
+
+    num_examples = len(session_data.X)
+    unique_labels, counts_labels = np.unique(
+        session_data.labels, return_counts=True, axis=0
+    )
+    num_labels = len(unique_labels)
+
+    # need to call every time, so that the data is shuffled inside each class
+    label_data = []
+    for label in unique_labels:
+        label_data.append(
+            SessionData(
+                X=session_data.X[session_data.labels == label],
+                Y=session_data.Y[session_data.labels == label],
+                labels=session_data.labels[session_data.labels == label],
+            )
+        )
+
+    data_idx = [0] * num_labels
+    num_data_cycles = [0] * num_labels
+    skipped = [False] * num_labels
+    new_X = []
+    new_Y = []
+    new_labels = []
+    while min(num_data_cycles) == 0:
+        if shuffle:
+            ids = np.random.permutation(num_labels)
+        else:
+            ids = range(num_labels)
+
+        for i in ids:
+            if num_data_cycles[i] > 0 and not skipped[i]:
+                skipped[i] = True
+                continue
+            else:
+                skipped[i] = False
+
+            num_i = int(counts_labels[i] / num_examples * batch_size) + 1
+
+            new_X.append(label_data[i].X[data_idx[i] : data_idx[i] + num_i])
```

Comment:
this results in the data being sorted within a batch (e.g. first all examples with label `A` then all with label `B`) - I guess that is ok, just wanted to confirm

Reply:
right, but within a batch we construct the loss by simple summation over batch examples, so order inside a batch shouldn't matter, no?